---
title: "Cancer Genomics"
author: "Matt Demelo"
date: "6/6/2019"
output:
  md_document:
    variant: markdown_github
always_allow_html: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Cancer Genomics: Sequence Comparison

Here we compare a wild-type and mutant cancer sequence to determine the sites of mutations within the mutant gene. First, let's load in the sequences:


```{r}
library(bio3d)
seqs <- read.fasta("lecture18_sequences.fa")
seqs

# We could check the alignment with this code:
# seqs <- seqaln(seqs)

```

Now we check the identity per aligned position, looking for regions with low identity and no gaps:

```{r}
## Calculate positional identity scores
ide <- conserv(seqs$ali, method="identity")
mutant.sites <- which(ide < 1) 

## Exclude gap possitions from analysis
gaps <- gap.inspect(seqs)
mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]

mutant.sites
```

These are our mutant sites within our alignments. Let's name these sequences appropriately:

```{r}
## Make a "names" label for our output sequences (one per mutant)
mutant.names <- paste0(seqs$ali["P53_wt",mutant.sites],
                       mutant.sites,
                       seqs$ali["P53_mutant",mutant.sites])

mutant.names

```

Now let's extract the sequence around each mutation:

```{r}
## Sequence positions surounding each mutant site
start.position <- mutant.sites - 8
end.position <-  mutant.sites + 8

# Blank matrix to store sub-sequences
store.seqs <- matrix("-", nrow=length(mutant.sites), ncol=17)
rownames(store.seqs) <- mutant.names

## Extract each sub-sequence
for(i in 1:length(mutant.sites)) {
  store.seqs[i,] <- seqs$ali["P53_mutant",start.position[i]:end.position[i]]
}

store.seqs
```

Now let's output these to a .fasta:

```{r}
write.fasta(seqs = store.seqs, ids = mutant.names, file = "subsequences.fa")
```


#### Side note: Making an Input Sequence

This isn't something we need to know, but here is how we would obtain a protooncogene sequence, generate it into a mutant form, and then output an aligned .fasta (i.e. prepare the file we used for this exercise)

```{r}
library(bio3d)

## Read KRas oncogene sequence from UniProt
wt <- get.seq("P01116")

## Here we make four mutants namely: G12V, Q22N, T74S and A130V
mutant <- wt
mutant$ali[ c(12,22,74,130)] <- c("V", "N", "S", "V")

write.fasta( seqbind(wt, mutant), ids=c("wt","mutant"), file="kras-sequences.fa")

```




```{r}
sessionInfo()
```







