---
title: "Lecture 15 - Pathway Analysis from RNA-Seq Results"
author: "Matt Demelo"
date: "5/21/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Section 1: RNA-Seq Analysis

The data for this exercise comes from GEO entry: GSE37704, associated with the following publication:

>Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. “Differential analysis of gene regulation at transcript resolution with RNA-seq”. Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

First, let's load in our datasets we are working with:

```{r}
# Meta Data from RNA-Seq
metaFile <- "https://bioboot.github.io/bimm143_W18/class-material/GSE37704_metadata.csv"

# Counts Data from RNA-Seq
countFile <- "https://bioboot.github.io/bimm143_W18/class-material/GSE37704_featurecounts.csv"

# Import metadata and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)
colata["SRR493366",]
```

Putting our count data into a useful form:

```{r}
# Let's put our count data into a data.frame
countData = read.csv(countFile, row.names = 1)
head(countData)
```

We want to remove the first "length" column in `countData`.

```{r}
countData <- as.matrix(countData[,-1])
head(countData)
```

Adding '-1' will remove one column from our matrix. Alternatively, we can just index that part of the matrix by adding "2:ncol(countData)" in the column specification.

However, if we were to run this multiple times, we start deleting count columns!

Let's remove rows with zero counts in all experiments (i.e. columns)

```{r}
# Filter count data where you have 0 read count across all samples.
countData = countData[rowSums(countData) > 0, ]
head(countData)

```

Alternatively, we can do nonzero.rows <- rowSums(countData) != 0, and then index countData using nonzero.rows


```{r}
nrow(countData)
```

Now it's time for DE analysis:

```{r}
# Load DESeq2 package
library(DESeq2)
```

```{r}
# DESeq2 analysis
dds <- DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds <- DESeq(dds)
```

Let's check our DESeq dataset:

```{r}
dds
```

Now let's look at the results

```{r}
res <- results(dds)
summary(res)
```

And now it's time to plot our data, to see what the data looks like:

```{r}

plot( res$log2FoldChange, -log(res$padj), 
      ylab="-Log(P-value)", xlab="Log2(FoldChange)", 
      main = "DESeq Analysis of GSE37704" )

abline(v = c(-2, 2), col = "green", lty = 2, lwd = 2)
abline(h = -log(0.01), col = "green", lty = 2, lwd = 2)
```

Now let's improve this plot

```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- (res$pvalue < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"


# The plot itself
plot( res$log2FoldChange, -log(res$padj), col= mycols, 
      xlab="Log2(FoldChange)", ylab="-Log(P-value)",
      main = "DESeq Analysis of GSE37704")

# Threshold lines
abline(v = c(-2, 2), col = "black", lty = 2, lwd = 2) # Vertical
abline(h = -log(0.01), col = "black", lty = 2, lwd = 2) # Horizontal
```

Now let's annotate the DESeq data:

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)

res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    keytype="ENSEMBL",
                    column="SYMBOL",
                    multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="GENENAME",
                    multiVals="first")

head(res, 10)
```

Finally, let's export our DESeq2 data to a .csv, after it's been ordered:

```{r}
res = res[order(res$pvalue),]
write.csv(res, file ="deseq_results.csv")
```

## Section 2: Pathways Analysis

First, let's load in and install the necessary packages. This was previously done, but here is the code:

```{r}
# source("http://bioconductor.org/biocLite.R")
# biocLite( c("pathview", "gage", "gageData") )
```

Loading the packages:

```{r}
library(pathview)
library(gage)
library(gageData)
```

Now we can start looking at some KEGG data:

```{r}

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

Let's make a simpler dataset to work with for our KEGG analysis. This will be a vector of 'foldchange' values. This is the format that our function wants.:
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

First we'll run our 'gage' pathway analysis:

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
attributes(keggres)
```

And observe the results:

```{r}
head(keggres$less)
```

Now let's gt a visual representation of our analysis by using the pathview package:

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

![](hsa04110.pathview.png)

Notice that many of the genes in our pathway analysis are "perturbed". This refers to the amount of expression, coloring indicating differentially expressed genes (Green indicates downregulation, Red indicates upregulation). Looking at the data, we can start to make sense of why some genes are upregulated in relation to others.

*Also, we can generate different file formats with pathview, such as a .pdf*:

```{r}
# A different PDF based output of the same data
# pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
# remove comment symbol above to make this code run.
```

Now let's take a look at the top 5 upregulated pathways:

```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

Now that we know this, let's pass these IDs through the pathview function to generate a figure of these pathways:

```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```

This generates several pathway picture files for each gene above.

![](hsa04640.pathview.png)
![](hsa04630.pathview.png)
![](hsa00140.pathview.png)
![](hsa04142.pathview.png)
![](hsa04330.pathview.png)

That's a lot of figures, but what if we just added one more?

![](http://3.bp.blogspot.com/_2W71Pm9UyGI/TE7khMlqVYI/AAAAAAAAAso/hMl2kGL5aXQ/s1600/100727_doginventsunderwatersunglassses.jpg)

Wait, this isn't data. How'd this get here?

## Section 3: Gene Ontologies (GO)

We can try a similar procedure as above, but with GO terms instead.

Let's start our GO analysis, focusing on Biological Processes (BP):

```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```

As we can see, GO gives us a pretty useful indication of what processes our genes are involved in. It's not as easy to digest visually as a pathway diagram, but it could be useful for further ordering of data, or just for having more identifiers to use in a future analysis.

## Section 4: Reactome Analysis

Reactome is a database which documents relationships of biologeical molecules and the processes/pathways they are part of. Let's try using it for our data.

We'll start by doing an over-representation enrichment analysis and pathway topology analysis using tools from Reactome. First, we need to generate an output of significantly expressed genes (alpha = 0.05) into a plain text file:

```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```

Now we simply upload the resultant file to Reactome, and it does the hard work for us. This is done at https://reactome.org/PathwayBrowser/#TOOL=AT

## Section 5: Online GO Analysis.

Genes from 'significant_genes.txt' were pasted into the GO annotation service at http://www.http://geneontology.org. GO search term was Biological Processes for Homo Sapiens.











